File: app/src/main/java/org/wen/gank/mvp/ViewDelegate.java
package org.wen.gank.mvp;

import android.os.Bundle;

/**
 * created by Jiahui.wen 2017-07-08
 */
abstract class ViewDelegate<V extends MvpView, P extends MvpPresenter<V>> {

    private P presenter;

    public void onCreate(Bundle saveInstance) {
        presenter = createPresenter();
        presenter.onCreate(saveInstance);
    }

    public void attachView(V view) {
        presenter.attachView(view);
    }

    protected void onSaveInstanceState(Bundle outState) {
        presenter.onSaveInstanceState(outState);
    }

    public void onDestroy() {
        presenter.onDestroy();
    }

    public void detachView() {
        presenter.detachView(false);
    }

    protected abstract P createPresenter();

    public P getPresenter() {
        return presenter;
    }
}

File: app/src/main/java/org/wen/gank/mvp/ViewDelegate.kt
package org.wen.gank.mvp

import android.os.Bundle

/**
 * created by Jiahui.wen 2017-07-08
 */
internal abstract class ViewDelegate<V : MvpView, P : MvpPresenter<V>> {

  lateinit var presenter: P

  fun onCreate(saveInstance: Bundle?) {
    presenter = createPresenter()
    presenter.onCreate(saveInstance)
  }

  fun attachView(view: V) {
    presenter.attachView(view)
  }

  fun onSaveInstanceState(outState: Bundle) {
    presenter.onSaveInstanceState(outState)
  }

  fun onDestroy() {
    presenter.onDestroy()
  }

  fun detachView() {
    presenter.detachView(false)
  }

  protected abstract fun createPresenter(): P
}
